### DISCUSSION TITLE
Published to the Mac App store

### DISCUSSION THREAD

**Discussion by fufufu3** [AvaloniaUI/Avalonia#7088] [Q&A]
After I read the official documents carefully, I still failed.  The generated PKG file cannot be uploaded.  I don't know if there are more detailed documents for some of these steps

**Comment by kekekeks** [comment]
@Mikolaytis has managed to publish Lunacy, there were some non-trivial requirements for the bundle structure, IIRC.

**Comment by Mikolaytis** [comment]
Hi. Did you read this part of the article? https://docs.avaloniaui.net/docs/distribution-publishing/macos#app-store-packaging
If everything is fine with this - provide more information about what's wrong. Errors, entitlements, scripts, etc. I will look into this and I hope I will be able to help you.

**Reply by fufufu3** [reply]
Yeah ,
thanks for your reply. I am not familiar with Avalonia yet. The article is useful,i success to make a dmg file to distribute my application,but now i want publish it on MacOS App Store.
i follow it then I ran into several difficulties.
1.the script in this article
"find "$APP_NAME/Contents/Frameworks/"|while read fname; do if [[ -f $fname ]]; then echo "[INFO] Signing $fname" codesign --force --sign "$APP_SIGNING_IDENTITY" "$fname" fi done"
can't work well when i used "ln -s MacOSfilename ToFrameworksFilename" to moving your .dylib files to Frameworks folder using relative symlinks.
2."main executable failed strict validation" when i signing Signing app executable

**Reply by Mikolaytis** [reply]
Download Lunacy from Mac App Store and check out our bundle. I think you are doing symlinks in a wrong way. MacOS folder should contain only executable and symlinks, other files should be in other folders

**Reply by fufufu3** [reply]
@Mikolaytis   i  got it wrong.
I restructured my bundle  ,but it failed when running
Error:
  An assembly specified in the application dependencies manifest (xxx.deps.json) was not found:
    package: 'runtimepack.Microsoft.NETCore.App.Runtime.osx-x64', version: '5.0.7'
    path: 'libSystem.IO.Compression.Native.dylib'
it means the relaive symlimnks is unusefull for my appliction???
How can I solve this problem

**Reply by Mikolaytis** [reply]
I think this error is solvable. Are you using net6? Re-check is symlinks valid. Also try to make a build without deps json. Single bundle should be without this AFAIK.

**Reply by fufufu3** [reply]
i am using net5. i will try it again with net6. thanks for your advice!!!

**Reply by Amenti** [reply]
I also can't get it to work even after reading all the documentation. I already fail at the publish step.

`dotnet publish -f net8.0-macos -c Release --self-contained true -p:PublishSingleFile=true -p:RuntimeIdentifier=osx-x64 -p:UseAppHost=true
`
always leads to "Invalid input specification: Must specify the host binary" even with a brand-new app with a quite simple Info.plist:

`/usr/local/share/dotnet/sdk/8.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(1038,5): error MSB4018: The "GenerateBundle" task failed unexpectedly. [/Users/username/RiderProjects/AvaloniaApplication4/AvaloniaApplication4/AvaloniaApplication4.csproj]
/usr/local/share/dotnet/sdk/8.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(1038,5): error MSB4018: System.ArgumentException: Invalid input specification: Must specify the host binary [/Users/username/RiderProjects/AvaloniaApplication4/AvaloniaApplication4/AvaloniaApplication4.csproj]
/usr/local/share/dotnet/sdk/8.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(1038,5): error MSB4018:    at Microsoft.NET.HostModel.Bundle.Bundler.GenerateBundle(IReadOnlyList`1 fileSpecs) [/Users/username/RiderProjects/AvaloniaApplication4/AvaloniaApplication4/AvaloniaApplication4.csproj]
/usr/local/share/dotnet/sdk/8.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(1038,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateBundle.ExecuteCore() [/Users/username/RiderProjects/AvaloniaApplication4/AvaloniaApplication4/AvaloniaApplication4.csproj]
/usr/local/share/dotnet/sdk/8.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(1038,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [/Users/username/RiderProjects/AvaloniaApplication4/AvaloniaApplication4/AvaloniaApplication4.csproj]
/usr/local/share/dotnet/sdk/8.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(1038,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/Users/username/RiderProjects/AvaloniaApplication4/AvaloniaApplication4/AvaloniaApplication4.csproj]
/usr/local/share/dotnet/sdk/8.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(1038,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/Users/username/RiderProjects/AvaloniaApplication4/AvaloniaApplication4/AvaloniaApplication4.csproj]
`
I would appreciate any idea into what I'm doing wrong here.

**Reply by Mikolaytis** [reply]
I use regular -f net8.0 also I dont use -p:UseAppHost=true

**Reply by Amenti** [reply]
Thanks for the info. Without -macos it works as well in my test app. UseAppHost seems to be a requirement for the -macos-Target as the documentation tells it.
My real app sadly needs the -macos or half of my code won't work since it uses AppKit APIs. How do you get around using those at all?

**Reply by maxkatz6** [reply]
If you already use "-macos" TFM, you probably don't need half of that documentation page. As SDK builds `.app` bundle for you, without need for all of these extra args. Just `dotnet publish` possibly with needed -f and -r params.

**Reply by maxkatz6** [reply]
It still doesn't pass sandboxing requirements in the store as built with SDK, so you would need to go through signing and sandboxing details on that documentation page.

**Reply by Amenti** [reply]
Thanks a lot. It works like a charm with the simpler call now.
